File: python/src/rivu_server_sdk/event_compactor.py
from __future__ import annotations

import json
import time
from copy import deepcopy
from dataclasses import dataclass
from typing import Any, Protocol

from .event_store import Envelope, InMemoryRingBufferEventStore
from .json_patch import JsonPatchError, apply_json_patch
from .snapshot_store import SqliteSnapshotStore


@dataclass(frozen=True)
class EventCompactorConfig:
    """Configuration for server-side flush/compaction on the storage boundary."""

    max_buffered_events: int | None = None
    max_buffered_bytes: int | None = None
    flush_interval_ms: int | None = None
    max_replay_events: int | None = None


@dataclass(frozen=True)
class EventCompactorMetrics:
    buffered_events: int
    buffered_bytes: int
    last_seq: int
    state_deltas_since_snapshot: int
    merged_text_chunks: int
    merged_tool_chunks: int
    snapshots_created: int
    flushes: int


class EventCompactor(Protocol):
    def push(self, envelope: Envelope) -> list[Envelope]: ...

    def flush(self) -> list[Envelope]: ...

    def metrics(self) -> EventCompactorMetrics: ...


def _now_ms() -> int:
    return int(time.time() * 1000)


def _estimate_envelope_bytes(env: Envelope) -> int:
    # Best-effort estimate for buffering/flush budgets.
    try:
        payload = json.dumps(env, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
        return len(payload)
    except Exception:
        return 0


def _get_str_field(event: dict[str, Any], key: str) -> str | None:
    v = event.get(key)
    if not isinstance(v, str):
        return None
    v = v.strip()
    return v if v else None


class DefaultEventCompactor:
    def __init__(self, config: EventCompactorConfig | None = None) -> None:
        self._config = config or EventCompactorConfig()
        self._buffer: list[Envelope] = []
        self._buffered_bytes: int = 0
        self._last_flush_at_ms: int = _now_ms()
        self._last_seq: int = 0
        self._shared_state: dict[str, Any] = {}
        self._state_deltas_since_snapshot: int = 0
        self._merged_text_chunks: int = 0
        self._merged_tool_chunks: int = 0
        self._snapshots_created: int = 0
        self._flushes: int = 0

    @property
    def config(self) -> EventCompactorConfig:
        return self._config

    def _should_flush_by_time(self) -> bool:
        interval = self._config.flush_interval_ms
        if interval is None or interval <= 0:
            return False
        if not self._buffer:
            return False
        return _now_ms() - self._last_flush_at_ms >= interval

    def _should_flush_by_budget(self) -> bool:
        max_events = self._config.max_buffered_events
        if max_events is not None and max_events > 0 and len(self._buffer) > max_events:
            return True
        max_bytes = self._config.max_buffered_bytes
        if max_bytes is not None and max_bytes > 0 and self._buffered_bytes > max_bytes:
            return True
        return False

    def _push_buffered(self, envelope: Envelope) -> None:
        self._buffered_bytes += _estimate_envelope_bytes(envelope)
        self._buffer.append(envelope)

    def _try_merge_text_chunk(self, envelope: Envelope) -> bool:
        event = envelope.get("event")
        if not isinstance(event, dict) or event.get("type") != "TEXT_MESSAGE_CHUNK":
            return False

        message_id = _get_str_field(event, "messageId")
        if message_id is None:
            return False
        role = _get_str_field(event, "role") or "assistant"
        delta = event.get("delta")
        if not isinstance(delta, str):
            delta = ""

        if not self._buffer:
            return False
        last = self._buffer[-1]
        last_event = last.get("event")
        if not isinstance(last_event, dict) or last_event.get("type") != "TEXT_MESSAGE_CHUNK":
            return False

        last_message_id = _get_str_field(last_event, "messageId")
        if last_message_id != message_id:
            return False
        last_role = _get_str_field(last_event, "role") or "assistant"
        if last_role != role:
            return False

        prev_bytes = _estimate_envelope_bytes(last)
        last_delta = last_event.get("delta")
        if not isinstance(last_delta, str):
            last_delta = ""
        last["seq"] = int(envelope["seq"])
        last_event["delta"] = f"{last_delta}{delta}"
        self._buffered_bytes = self._buffered_bytes - prev_bytes + _estimate_envelope_bytes(last)
        self._merged_text_chunks += 1
        return True

    def _try_merge_tool_chunk(self, envelope: Envelope) -> bool:
        event = envelope.get("event")
        if not isinstance(event, dict) or event.get("type") != "TOOL_CALL_CHUNK":
            return False

        tool_call_id = _get_str_field(event, "toolCallId")
        if tool_call_id is None:
            return False
        delta = event.get("delta")
        if not isinstance(delta, str):
            delta = ""

        name = _get_str_field(event, "toolCallName")
        parent = _get_str_field(event, "parentMessageId")

        if not self._buffer:
            return False
        last = self._buffer[-1]
        last_event = last.get("event")
        if not isinstance(last_event, dict) or last_event.get("type") != "TOOL_CALL_CHUNK":
            return False

        last_tool_call_id = _get_str_field(last_event, "toolCallId")
        if last_tool_call_id != tool_call_id:
            return False

        last_name = _get_str_field(last_event, "toolCallName")
        if last_name is not None and name is not None and last_name != name:
            return False
        last_parent = _get_str_field(last_event, "parentMessageId")
        if last_parent is not None and parent is not None and last_parent != parent:
            return False

        prev_bytes = _estimate_envelope_bytes(last)
        last_delta = last_event.get("delta")
        if not isinstance(last_delta, str):
            last_delta = ""

        last["seq"] = int(envelope["seq"])
        last_event["delta"] = f"{last_delta}{delta}"
        if last_name is None and name is not None:
            last_event["toolCallName"] = name
        if last_parent is None and parent is not None:
            last_event["parentMessageId"] = parent

        self._buffered_bytes = self._buffered_bytes - prev_bytes + _estimate_envelope_bytes(last)
        self._merged_tool_chunks += 1
        return True

    def _on_state_snapshot(self, event: dict[str, Any]) -> None:
        snapshot = event.get("snapshot")
        if not isinstance(snapshot, dict):
            raise ValueError("STATE_SNAPSHOT.snapshot must be an object")
        self._shared_state = deepcopy(snapshot)
        self._state_deltas_since_snapshot = 0

    def _on_state_delta(self, event: dict[str, Any]) -> None:
        delta = event.get("delta")
        if not isinstance(delta, list):
            raise ValueError("STATE_DELTA.delta must be an array")
        try:
            next_state = apply_json_patch(self._shared_state, delta)  # type: ignore[arg-type]
        except JsonPatchError as e:
            raise ValueError("apply_json_patch failed") from e
        if not isinstance(next_state, dict):
            raise ValueError("patch result must be an object")
        self._shared_state = next_state
        self._state_deltas_since_snapshot += 1

    def _maybe_truncate_patch_chain(self, *, seq: int) -> None:
        max_replay_events = self._config.max_replay_events
        if max_replay_events is None or max_replay_events <= 0:
            return
        if self._state_deltas_since_snapshot <= max_replay_events:
            return

        last_snapshot_seq = 0
        for env in reversed(self._buffer):
            ev = env.get("event")
            if isinstance(ev, dict) and ev.get("type") == "STATE_SNAPSHOT":
                last_snapshot_seq = int(env.get("seq", 0) or 0)
                break

        self._buffer = [
            env
            for env in self._buffer
            if not (isinstance(env.get("event"), dict) and env["event"].get("type") == "STATE_DELTA" and int(env.get("seq", 0)) > last_snapshot_seq)
        ]
        self._buffer.append({"seq": int(seq), "event": {"type": "STATE_SNAPSHOT", "snapshot": deepcopy(self._shared_state)}})
        self._buffered_bytes = sum(_estimate_envelope_bytes(e) for e in self._buffer)
        self._state_deltas_since_snapshot = 0
        self._snapshots_created += 1

    def push(self, envelope: Envelope) -> list[Envelope]:
        seq = envelope.get("seq")
        if not isinstance(seq, int) or seq <= 0:
            raise ValueError("envelope.seq must be a positive int")

        flushed: list[Envelope] = []
        if self._should_flush_by_time():
            flushed.extend(self.flush())

        self._last_seq = max(self._last_seq, seq)

        if self._try_merge_text_chunk(envelope):
            pass
        elif self._try_merge_tool_chunk(envelope):
            pass
        else:
            event = envelope.get("event")
            self._push_buffered({"seq": int(seq), "event": deepcopy(event)})

        event = envelope.get("event")
        if isinstance(event, dict):
            event_type = event.get("type")
            if event_type == "STATE_SNAPSHOT":
                self._on_state_snapshot(event)
            elif event_type == "STATE_DELTA":
                self._on_state_delta(event)
                self._maybe_truncate_patch_chain(seq=seq)

        if self._should_flush_by_budget():
            flushed.extend(self.flush())

        return flushed

    def flush(self) -> list[Envelope]:
        if not self._buffer:
            return []
        self._flushes += 1
        self._last_flush_at_ms = _now_ms()
        out = self._buffer
        self._buffer = []
        self._buffered_bytes = 0
        return out

    def metrics(self) -> EventCompactorMetrics:
        return EventCompactorMetrics(
            buffered_events=len(self._buffer),
            buffered_bytes=self._buffered_bytes,
            last_seq=self._last_seq,
            state_deltas_since_snapshot=self._state_deltas_since_snapshot,
            merged_text_chunks=self._merged_text_chunks,
            merged_tool_chunks=self._merged_tool_chunks,
            snapshots_created=self._snapshots_created,
            flushes=self._flushes,
        )


class CompactingEventSink:
    """Optional helper that persists compactor output into the default stores."""

    def __init__(
        self,
        *,
        thread_id: str,
        event_store: InMemoryRingBufferEventStore,
        snapshot_store: SqliteSnapshotStore,
        compactor: DefaultEventCompactor | None = None,
    ) -> None:
        if not thread_id.strip():
            raise ValueError("thread_id must be non-empty")
        self._thread_id = thread_id
        self._event_store = event_store
        self._snapshot_store = snapshot_store
        self._compactor = compactor or DefaultEventCompactor()

    def push(self, envelope: Envelope) -> None:
        out = self._compactor.push(envelope)
        self._persist(out)

    def flush(self) -> None:
        out = self._compactor.flush()
        self._persist(out)

    def metrics(self) -> EventCompactorMetrics:
        return self._compactor.metrics()

    def _persist(self, envelopes: list[Envelope]) -> None:
        for env in envelopes:
            event = env.get("event")
            if isinstance(event, dict) and event.get("type") == "STATE_SNAPSHOT":
                snapshot = event.get("snapshot")
                if not isinstance(snapshot, dict):
                    raise ValueError("STATE_SNAPSHOT.snapshot must be an object")
                self._snapshot_store.put(thread_id=self._thread_id, seq=int(env["seq"]), shared_state=snapshot)
            self._event_store.append(self._thread_id, env)

File: crates/rivu-server-sdk/src/event_compactor.rs
use std::time::{SystemTime, UNIX_EPOCH};

use serde_json::{json, Value};

use crate::event_store::Envelope;
use crate::snapshot_store::SqliteSnapshotStore;
use crate::ui_spec::apply_json_patch;
use crate::{EventStoreError, InMemoryRingBufferEventStore, SnapshotStoreError};

#[derive(Debug, Clone, Copy, Default)]
pub struct EventCompactorConfig {
    /// Flush buffered envelopes when more than this many are buffered.
    ///
    /// Use `None` to disable.
    pub max_buffered_events: Option<usize>,

    /// Flush buffered envelopes when buffered JSON bytes exceed this budget.
    ///
    /// Use `None` to disable.
    pub max_buffered_bytes: Option<usize>,

    /// Flush buffered envelopes when more than this many milliseconds have elapsed since the last flush.
    ///
    /// Use `None` to disable.
    pub flush_interval_ms: Option<u64>,

    /// When the number of `STATE_DELTA` events since the last snapshot exceeds this budget, emit a
    /// `STATE_SNAPSHOT` to truncate the patch chain.
    ///
    /// Use `None` to disable.
    pub max_replay_events: Option<usize>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct EventCompactorMetrics {
    pub buffered_events: usize,
    pub buffered_bytes: usize,
    pub last_seq: u64,
    pub state_deltas_since_snapshot: usize,
    pub merged_text_chunks: u64,
    pub merged_tool_chunks: u64,
    pub snapshots_created: u64,
    pub flushes: u64,
}

#[derive(thiserror::Error, Debug)]
pub enum EventCompactorError {
    #[error("invalid envelope seq (must be > 0)")]
    InvalidSeq,

    #[error("STATE_SNAPSHOT.snapshot must be a json object")]
    SnapshotNotObject,

    #[error("STATE_DELTA.delta must be a json array")]
    DeltaNotArray,

    #[error("failed to apply json patch")]
    PatchApplyFailed,

    #[error(transparent)]
    EventStore(#[from] EventStoreError),

    #[error(transparent)]
    SnapshotStore(#[from] SnapshotStoreError),
}

pub trait EventCompactor {
    fn push(&mut self, envelope: Envelope) -> Result<Vec<Envelope>, EventCompactorError>;
    fn flush(&mut self) -> Vec<Envelope>;
    fn metrics(&self) -> EventCompactorMetrics;
}

#[derive(Debug, Clone)]
pub struct DefaultEventCompactor {
    config: EventCompactorConfig,
    buffer: Vec<Envelope>,
    buffered_bytes: usize,
    last_flush_at_ms: u64,
    last_seq: u64,
    shared_state: Value,
    state_deltas_since_snapshot: usize,
    merged_text_chunks: u64,
    merged_tool_chunks: u64,
    snapshots_created: u64,
    flushes: u64,
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn estimate_envelope_bytes(env: &Envelope) -> usize {
    // Best-effort estimate for buffering/flush budgets.
    serde_json::to_vec(env).map(|v| v.len()).unwrap_or(0)
}

fn get_string_field<'a>(event: &'a Value, key: &str) -> Option<&'a str> {
    event.get(key).and_then(Value::as_str).map(|s| s.trim()).filter(|s| !s.is_empty())
}

impl DefaultEventCompactor {
    pub fn new(config: EventCompactorConfig) -> Self {
        Self {
            config,
            buffer: vec![],
            buffered_bytes: 0,
            last_flush_at_ms: now_ms(),
            last_seq: 0,
            shared_state: Value::Object(serde_json::Map::new()),
            state_deltas_since_snapshot: 0,
            merged_text_chunks: 0,
            merged_tool_chunks: 0,
            snapshots_created: 0,
            flushes: 0,
        }
    }

    pub fn config(&self) -> EventCompactorConfig {
        self.config
    }

    fn should_flush_by_time(&self) -> bool {
        let Some(interval) = self.config.flush_interval_ms else {
            return false;
        };
        if self.buffer.is_empty() {
            return false;
        }
        now_ms().saturating_sub(self.last_flush_at_ms) >= interval
    }

    fn should_flush_by_budget(&self) -> bool {
        if let Some(max) = self.config.max_buffered_events {
            if self.buffer.len() > max {
                return true;
            }
        }
        if let Some(max) = self.config.max_buffered_bytes {
            if self.buffered_bytes > max {
                return true;
            }
        }
        false
    }

    fn push_buffered(&mut self, envelope: Envelope) {
        self.buffered_bytes += estimate_envelope_bytes(&envelope);
        self.buffer.push(envelope);
    }

    fn try_merge_text_chunk(&mut self, envelope: &Envelope) -> bool {
        let event = &envelope.event;
        if event.get("type").and_then(Value::as_str) != Some("TEXT_MESSAGE_CHUNK") {
            return false;
        }
        let message_id = get_string_field(event, "messageId");
        if message_id.is_none() {
            return false;
        }
        let role = get_string_field(event, "role").unwrap_or("assistant");
        let delta = event.get("delta").and_then(Value::as_str).unwrap_or("");

        let Some(last) = self.buffer.last_mut() else {
            return false;
        };
        if last.event.get("type").and_then(Value::as_str) != Some("TEXT_MESSAGE_CHUNK") {
            return false;
        }
        let last_message_id = get_string_field(&last.event, "messageId");
        if last_message_id != message_id {
            return false;
        }
        let last_role = get_string_field(&last.event, "role").unwrap_or("assistant");
        if last_role != role {
            return false;
        }

        let prev_bytes = estimate_envelope_bytes(last);
        let last_delta = last.event.get("delta").and_then(Value::as_str).unwrap_or("");
        let combined = format!("{last_delta}{delta}");
        last.seq = envelope.seq;
        last.event["delta"] = Value::String(combined);
        self.buffered_bytes = self.buffered_bytes.saturating_sub(prev_bytes) + estimate_envelope_bytes(last);
        self.merged_text_chunks += 1;
        true
    }

    fn try_merge_tool_chunk(&mut self, envelope: &Envelope) -> bool {
        let event = &envelope.event;
        if event.get("type").and_then(Value::as_str) != Some("TOOL_CALL_CHUNK") {
            return false;
        }
        let tool_call_id = get_string_field(event, "toolCallId");
        if tool_call_id.is_none() {
            return false;
        }
        let delta = event.get("delta").and_then(Value::as_str).unwrap_or("");
        let name = get_string_field(event, "toolCallName").map(|s| s.to_string());
        let parent = get_string_field(event, "parentMessageId").map(|s| s.to_string());

        let Some(last) = self.buffer.last_mut() else {
            return false;
        };
        if last.event.get("type").and_then(Value::as_str) != Some("TOOL_CALL_CHUNK") {
            return false;
        }
        let last_tool_call_id = get_string_field(&last.event, "toolCallId");
        if last_tool_call_id != tool_call_id {
            return false;
        }

        let last_name = get_string_field(&last.event, "toolCallName").map(|s| s.to_string());
        if let (Some(a), Some(b)) = (&last_name, &name) {
            if a != b {
                return false;
            }
        }
        let last_parent = get_string_field(&last.event, "parentMessageId").map(|s| s.to_string());
        if let (Some(a), Some(b)) = (&last_parent, &parent) {
            if a != b {
                return false;
            }
        }

        let prev_bytes = estimate_envelope_bytes(last);
        let last_delta = last.event.get("delta").and_then(Value::as_str).unwrap_or("");
        let combined = format!("{last_delta}{delta}");
        last.seq = envelope.seq;
        last.event["delta"] = Value::String(combined);
        if last_name.is_none() {
            if let Some(name) = &name {
                last.event["toolCallName"] = Value::String(name.clone());
            }
        }
        if last_parent.is_none() {
            if let Some(parent) = &parent {
                last.event["parentMessageId"] = Value::String(parent.clone());
            }
        }
        self.buffered_bytes = self.buffered_bytes.saturating_sub(prev_bytes) + estimate_envelope_bytes(last);
        self.merged_tool_chunks += 1;
        true
    }

    fn on_state_snapshot(&mut self, envelope: &Envelope) -> Result<(), EventCompactorError> {
        let snapshot = envelope.event.get("snapshot").cloned().unwrap_or(Value::Null);
        if !snapshot.is_object() {
            return Err(EventCompactorError::SnapshotNotObject);
        }
        self.shared_state = snapshot;
        self.state_deltas_since_snapshot = 0;
        Ok(())
    }

    fn on_state_delta(&mut self, envelope: &Envelope) -> Result<(), EventCompactorError> {
        let delta = envelope
            .event
            .get("delta")
            .and_then(Value::as_array)
            .cloned()
            .ok_or(EventCompactorError::DeltaNotArray)?;
        let next = apply_json_patch(&self.shared_state, &delta).map_err(|_| EventCompactorError::PatchApplyFailed)?;
        if !next.is_object() {
            return Err(EventCompactorError::PatchApplyFailed);
        }
        self.shared_state = next;
        self.state_deltas_since_snapshot += 1;
        Ok(())
    }

    fn maybe_truncate_patch_chain(&mut self, seq: u64) {
        let Some(max) = self.config.max_replay_events else {
            return;
        };
        if self.state_deltas_since_snapshot <= max {
            return;
        }

        // Keep snapshots; drop buffered STATE_DELTA events after the last snapshot.
        let last_snapshot_seq = self
            .buffer
            .iter()
            .rev()
            .find(|env| env.event.get("type").and_then(Value::as_str) == Some("STATE_SNAPSHOT"))
            .map(|env| env.seq)
            .unwrap_or(0);

        self.buffer.retain(|env| {
            let t = env.event.get("type").and_then(Value::as_str).unwrap_or("");
            t != "STATE_DELTA" || env.seq <= last_snapshot_seq
        });

        self.buffer.push(Envelope {
            seq,
            event: json!({ "type": "STATE_SNAPSHOT", "snapshot": self.shared_state.clone() }),
        });

        self.buffered_bytes = self.buffer.iter().map(estimate_envelope_bytes).sum();
        self.state_deltas_since_snapshot = 0;
        self.snapshots_created += 1;
    }
}

impl EventCompactor for DefaultEventCompactor {
    fn push(&mut self, envelope: Envelope) -> Result<Vec<Envelope>, EventCompactorError> {
        if envelope.seq == 0 {
            return Err(EventCompactorError::InvalidSeq);
        }

        let mut flushed: Vec<Envelope> = vec![];

        if self.should_flush_by_time() {
            flushed.extend(self.flush());
        }

        self.last_seq = envelope.seq.max(self.last_seq);

        if self.try_merge_text_chunk(&envelope) {
            // merged
        } else if self.try_merge_tool_chunk(&envelope) {
            // merged
        } else {
            self.push_buffered(envelope.clone());
        }

        let event_type = envelope.event.get("type").and_then(Value::as_str).unwrap_or("");
        if event_type == "STATE_SNAPSHOT" {
            self.on_state_snapshot(&envelope)?;
        } else if event_type == "STATE_DELTA" {
            self.on_state_delta(&envelope)?;
            self.maybe_truncate_patch_chain(envelope.seq);
        }

        if self.should_flush_by_budget() {
            flushed.extend(self.flush());
        }

        Ok(flushed)
    }

    fn flush(&mut self) -> Vec<Envelope> {
        if self.buffer.is_empty() {
            return vec![];
        }
        self.flushes += 1;
        self.last_flush_at_ms = now_ms();
        self.buffered_bytes = 0;
        std::mem::take(&mut self.buffer)
    }

    fn metrics(&self) -> EventCompactorMetrics {
        EventCompactorMetrics {
            buffered_events: self.buffer.len(),
            buffered_bytes: self.buffered_bytes,
            last_seq: self.last_seq,
            state_deltas_since_snapshot: self.state_deltas_since_snapshot,
            merged_text_chunks: self.merged_text_chunks,
            merged_tool_chunks: self.merged_tool_chunks,
            snapshots_created: self.snapshots_created,
            flushes: self.flushes,
        }
    }
}

/// Optional helper that persists compactor output into the default stores:
/// - compacted envelopes → `InMemoryRingBufferEventStore`
/// - emitted `STATE_SNAPSHOT` → `SqliteSnapshotStore`
pub struct CompactingEventSink<'a> {
    thread_id: &'a str,
    event_store: &'a InMemoryRingBufferEventStore,
    snapshot_store: &'a SqliteSnapshotStore,
    compactor: DefaultEventCompactor,
}

impl<'a> CompactingEventSink<'a> {
    pub fn new(
        thread_id: &'a str,
        event_store: &'a InMemoryRingBufferEventStore,
        snapshot_store: &'a SqliteSnapshotStore,
        compactor: DefaultEventCompactor,
    ) -> Self {
        Self {
            thread_id,
            event_store,
            snapshot_store,
            compactor,
        }
    }

    pub fn push(&mut self, envelope: Envelope) -> Result<(), EventCompactorError> {
        let out = self.compactor.push(envelope)?;
        self.persist(out)?;
        Ok(())
    }

    pub fn flush(&mut self) -> Result<(), EventCompactorError> {
        let out = self.compactor.flush();
        self.persist(out)?;
        Ok(())
    }

    pub fn metrics(&self) -> EventCompactorMetrics {
        self.compactor.metrics()
    }

    fn persist(&self, envelopes: Vec<Envelope>) -> Result<(), EventCompactorError> {
        for env in envelopes {
            if env.event.get("type").and_then(Value::as_str) == Some("STATE_SNAPSHOT") {
                let snapshot = env.event.get("snapshot").cloned().unwrap_or(Value::Null);
                if !snapshot.is_object() {
                    return Err(EventCompactorError::SnapshotNotObject);
                }
                self.snapshot_store.put(self.thread_id, env.seq, &snapshot)?;
            }
            self.event_store.append(self.thread_id, env)?;
        }
        Ok(())
    }
}
